init.lua

```diff
@@ -820,6 +824,34 @@ function Form:set_as_inventory_for(player, ctx)
     player:set_inventory_formspec(fs)
 end
 
+-- Declared here to be accessable by render_to_formspec_string
+local fs_process_events
+
+-- Returns a tuple of string and function
```

Comment:
What about `size[]`?

Reply:
Isn't this size information provided by the root element returned from the callback to `flow.make_gui`?

Reply:
The `size[]` is returned in the formspec but if you want to embed it in an inventory mod then it would need to be removed.

Reply:
Is this true for `formspec_version[]` as well?

Reply:
Yes. I'm not sure what the best way to return them would be

Reply:
From testing, this doesn't seem to actually cause any problems to have these, but if we'd like to stay on spec, conditionally removing `size[]` looks easy (can be done the same way `no_prepend[]` is done), and after render, `formspec_version[]` can be removed with RegExps.

Should we even worry about this, since it works without removing them? Is there a better way to remove `formspec_version[]`?

Reply:
> Should we even worry about this, since it works without removing them?

I'd rather they be returned separately

> Is there a better way to remove formspec_version[]?

Maybe something like this?

```lua
local tree, form_info = form:_render(...)
local header = formspec_ast.unparse({
    formspec_version = tree.formspec_version,
    table.remove(tree, 1)  -- Remove the size[] from the tree
})
tree.formspec_version = nil
local fs = formspec_ast.unparse(tree)
```

Reply:
I figured I'd do this via an optional boolean as the third arg. Some apis, like sfinv, allow you to just set the whole formspec (`sfinv.register_page`) and some only embed, so both are needed.

Reply:
> I figured I'd do this via an optional boolean as the third arg. Some apis, like sfinv, allow you to just set the whole formspec (`sfinv.register_page`) and some only embed, so both are needed.

`sfinv.make_formspec` has a separate argument for the size element:

```lua
sfinv.register_page("test:page", {
    title = "Test page",
    get = function(self, player, context)
        local fs, on_event, w, h = form:render_to_formspec_string(...)
        context.on_event = on_event
        return sfinv.make_formspec(player, context, fs, true, ("formspec_version[6]size[%s,%s]"):format(w, h))
    end,
    on_player_receive_fields = function(self, player, context, fields)
        if not context.on_event or context.on_event(fields) then
            sfinv.set_player_inventory_formspec(player, context)
        end
    end,
})
```

Reply:
Right. Thought I'd clear a miscommunication with how I actually implemented it:

```lua
sfinv.register_page("test:page", {
    title = "Test page",
    get = function(self, player, context)
        -- Info is only present when the third arg is truthy (embed mode)
        local fs, on_event, info = form:render_to_formspec_string(..., true)
        context.test_page_event = on_event
        return sfinv.make_formspec(player, context, fs, true, ("formspec_version[%s]size[%s,%s]"):format(info.formspec_version, info.w, info h))
    end,
    on_player_receive_fields = function(self, player, context, fields)
        if not context.test_page_event or context.test_page_event(fields) then
            sfinv.set_player_inventory_formspec(player, context)
        end
    end,
})
```

It's an argument to `Flow:render_to_formspec_string` I was talking about. I was trying to point out that sometimes we _do_ want the size and version, but other times we don't, so it should be an arg.

Reply:
I wasn't sure where `render_to_formspec_string` could be used when `embedded` is false, I guess it could be used if you want to create a sfinv page without the navigation bar.

Reply:
Sorry to reopen. It occurs to me that there's a possibility we want embedded mode to be the default. What do you think?

Reply:
That sounds like a good idea.